fw/http_frame.c

```diff
@@ -2445,76 +2456,47 @@ do {									\
 #undef ADJUST_AVAILABLE_SIZE
 }
 
-static int
-__tfw_h2_make_frames(TfwH2Ctx *ctx, struct rb_node *node,
-		     unsigned long *cwnd_awail, unsigned int mss)
-{
-	int r;
-
-	if (node && (*cwnd_awail > FRAME_HEADER_SIZE)) {
-		TfwStream *stream = rb_entry(node, TfwStream, node);
-		if (stream->xmit.skb_head) {
-			BUG_ON(!stream->xmit.h_len && !stream->xmit.b_len);
-			if ((r = tfw_h2_make_frames_for_stream(ctx, stream,
-							       cwnd_awail,
-							       mss)))
-				return r;
-		}
-
-		if ((r = __tfw_h2_make_frames(ctx, node->rb_left, cwnd_awail,
-					      mss)))
-			return r;
-		if ((r = __tfw_h2_make_frames(ctx, node->rb_right, cwnd_awail,
-					      mss)))
-			return r;
-	}
-
-	return 0;
-}
-
-/**
- * TODO0 #1196 A very inefficient functiom, should be reworked when an efficient
- * stream prioritization data structure will be implemented.
- */
-static void
-__tfw_h2_check_available_data(struct rb_node *node, bool *data_is_available)
-{
-	if (node) {
-		TfwStream *stream = rb_entry(node, TfwStream, node);
-		if (stream->xmit.skb_head) {
-			*data_is_available = true;
-			return;
-		}
-		__tfw_h2_check_available_data(node->rb_left, data_is_available);
-		__tfw_h2_check_available_data(node->rb_right, data_is_available);
-	}
-}
-
 int
 tfw_h2_make_frames(TfwH2Ctx *ctx, unsigned long cwnd_awail, unsigned int mss,
 		   bool *data_is_available)
 {
 	TfwStreamSched *sched = &ctx->sched;
-	struct rb_node *root = sched->streams.rb_node;
-	int r;
+	TfwStreamSchedEntry *parent;
+	TfwStream *stream;
+	int r = 0;
 
 	BUG_ON(mss <= FRAME_HEADER_SIZE);
 
-	if (unlikely(r = __tfw_h2_make_frames(ctx, root, &cwnd_awail,
-					      mss - FRAME_HEADER_SIZE)))
-		return r;
+	while (tfw_h2_stream_sched_is_active(&sched->root)
+	       && cwnd_awail >= FRAME_HEADER_SIZE && ctx->rem_wnd && !r)
+	{
+		stream = tfw_h2_sched_stream_dequeue(&sched->root, &parent);
+		/*
+		 * If root scheduler is active we always can find
+		 * active stream.
+		 */
+		BUG_ON(!stream);
+		r = tfw_h2_make_frames_for_stream(ctx, stream, &cwnd_awail,
+						  mss);
+		tfw_h2_sched_stream_enqueue(stream, parent);
+	}
 
-	__tfw_h2_check_available_data(root, data_is_available);
-	return 0;
+	*data_is_available = tfw_h2_stream_sched_is_active(&sched->root);
+	return r;
 }
 
 static void
 __tfw_h2_purge_stream_send_queue(struct rb_node *node)
 {
 	if (node) {
 		TfwStream *stream = rb_entry(node, TfwStream, node);
-		if (stream->xmit.skb_head)
+		TfwStreamSchedEntry *parent = stream->sched.parent;
+
+		if (stream->xmit.skb_head) {
+			tfw_h2_stream_schred_remove(stream);
 			tfw_h2_stream_purge_send_queue(stream);
+			tfw_h2_sched_stream_enqueue(stream, parent);
+		}
 		__tfw_h2_purge_stream_send_queue(node->rb_left);
```

Comment:
Is recursion is available for us? Kernel stack is small, so can we use recursion?